src/x11.rs

```diff
@@ -187,138 +207,205 @@ impl X11Impl {
+                seg: None,
+                done_processing: None,
+            })
+        } else {
+            // SHM is not available.
+            Buffer::Wire(Vec::new())
         };
 
         Ok(Self {
             display,
             window,
             gc,
             depth: geometry_reply.depth,
-            shm: shm_info,
+            buffer,
+            width: 0,
+            height: 0,
         })
     }
 
-    pub(crate) unsafe fn set_buffer(&mut self, buffer: &[u32], width: u16, height: u16) {
-        // Draw the image to the buffer.
-        let result = unsafe { self.set_buffer_shm(buffer, width, height) }.and_then(|had_shm| {
-            if had_shm {
-                Ok(())
-            } else {
-                log::debug!("Falling back to non-SHM method");
-                self.set_buffer_fallback(buffer, width, height)
-            }
-        });
-
-        if let Err(e) = result {
-            log::error!("Failed to draw image to window: {}", e);
+    /// Resize the internal buffer to the given width and height.
+    pub(crate) fn resize(&mut self, width: u32, height: u32) -> Result<(), SoftBufferError> {
+        log::trace!(
+            "resize: window={:X}, size={}x{}",
+            self.window,
+            width,
+            height
+        );
+
+        // Width and height should fit in u16.
+        let width: u16 = width
+            .try_into()
+            .or(Err(SoftBufferError::SizeOutOfRange { width, height }))?;
+        let height: u16 = height.try_into().or(Err(SoftBufferError::SizeOutOfRange {
+            width: width.into(),
+            height,
+        }))?;
+
+        if width != self.width || height != self.height {
+            self.buffer
+                .resize(&self.display.connection, width, height)
+                .map_err(|err| {
+                    SoftBufferError::PlatformError(
+                        Some("Failed to resize X11 buffer".to_string()),
+                        Some(Box::new(err)),
+                    )
+                })?;
+
+            // We successfully resized the buffer.
+            self.width = width;
+            self.height = height;
         }
+
+        Ok(())
     }
 
-    /// Put the given buffer into the window using the SHM method.
-    ///
-    /// Returns `false` if SHM is not available.
-    ///
-    /// # Safety
-    ///
-    /// The buffer's length must be `width * height`.
-    unsafe fn set_buffer_shm(
-        &mut self,
-        buffer: &[u32],
-        width: u16,
-        height: u16,
-    ) -> Result<bool, PushBufferError> {
-        let shm_info = match self.shm {
-            Some(ref mut info) => info,
-            None => return Ok(false),
-        };
+    /// Get a mutable reference to the buffer.
+    pub(crate) fn buffer_mut(&mut self) -> Result<BufferImpl, SoftBufferError> {
+        log::trace!("buffer_mut: window={:X}", self.window);
+
+        Ok(BufferImpl(util::BorrowStack::new(self, |surface| {
```

Comment:
I think that we can avoid this antipattern. If we call `finish_wait` before we return the buffer, it's guaranteed that Derefing the buffer will succeed.

Reply:
Really need some kind of benchmark to compare things like whether storing the slice helps.

The `BorrowStack` code should be safe, but admittedly is someone unusual and perhaps a bit confusing to read. Hopefully at some point Rust will have a better way to deal with this and self-referential structs.